### PR TITLE
Add support for server-time

### PR DESCRIPTION
**Pull request by progval**
https://ircv3.net/specs/extensions/server-time

This is used by servers and bouncers when replaying history.

This commit adds a new ~~`const GDateTime *time`~~ `const SircMessageContext*`  parameter to all callback functions,
containing the time specified by the server if any, and ~~NULL~~ the current time otherwise.

~~I wrote this a while ago. In retrospect, I'm not sure it's a good solution to add a new
argument just for this; because such arguments would just accumulate with more
context info if Srain gets support for other specifications in the future (eg.
[message-specific display names](https://github.com/ircv3/ircv3-specifications/pull/452),
[bot message tags](https://ircv3.net/specs/extensions/bot-mode),
[message IDs](https://ircv3.net/specs/extensions/message-ids) and [replies](https://ircv3.net/specs/client-tags/reply), etc.)~~

~~So I wonder if I should rewrite this PR to pass some sort of structure with
all the context info instead. Possibly pass all the tags directly.
What do you think?~~

EDIT: I replaced it with a structure

### CODE REVIEW

**Comment by SilverRainZ** [comment]
Can use use a GVariantDict for such params?

**Comment by progval** [comment]
Hmm, the possible items should be known at compile time. Unless you want to add support for adding items via plugins in the future?

**Comment by SilverRainZ** [comment]
If all items are known at compile time, how about adding a  struct?

**Comment by progval** [comment]
Hey, I finally found some free time to work on this. I rebased the PR on master and rewrote it to use an opaque structure.

**Comment by progval** [comment]
By the way, the way I use to test this is to connect to irc.ergo.chat, join an existing room with some messages, and type `/quote history #channelname`.

It looks like this:

screenshot-2021-12-23_13-19-09

(HistServ is a pseudo-client used by Ergo to announce non-messages, when clients don't support the [`draft/event-playback`](https://ircv3.net/specs/extensions/chathistory) capability)

File: src/sirc/sirc.c [hunk 1]
```diff
@@ -359,31 +359,46 @@ static void on_disconnect_ready(GObject *obj, GAsyncResult *result, gpointer use
...
     LOG_FR("Connected");
+    SircMessageContext *context = sirc_message_context_new(NULL);
 
     sirc->stream = stream;
     sirc_recv(sirc);
 
-    g_return_if_fail(sirc->events->connect);
-    sirc->events->connect(sirc, "CONNECT");
+    if (!sirc->events->connect) {
+        sirc_message_context_free(context);
```
Comment (SilverRainZ) [inline src/sirc/sirc.c:368]: How about using [g_auto](http://www.manpagez.com/html/glib/glib-2.46.2/glib-Miscellaneous-Macros.php#g-auto) for freeing context?

File: src/sirc/sirc_context.c [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+};
+
+SircMessageContext* sirc_message_context_new(GDateTime* time) {
+    SircMessageContext *context;
+
+    if (!time) {
+        time = g_date_time_new_now_local();
+    }
+
+    g_return_val_if_fail(time, NULL);
```
Comment (SilverRainZ) [inline src/sirc/sirc_context.c:33]: This check can be omitted, I think.

File: src/sirc/sirc_context.c [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+    context = g_malloc0(sizeof(SircMessageContext));
+    g_return_val_if_fail(context, NULL);
```
Comment (SilverRainZ) [inline src/sirc/sirc_context.c:36]: Can be omitted too, GTK application won't works when malloc failed.

File: src/sirc/sirc_event_hdr.c [hunk 4]
```diff
@@ -34,9 +34,36 @@
...
-static void sirc_ctcp_event_hdr(SircSession *sirc, SircMessage *imsg);
+static void sirc_ctcp_event_hdr(SircSession *sirc, SircMessage *imsg, const SircMessageContext *context);
+
+void _sirc_event_hdr(SircSession *sirc, SircMessage *imsg, const SircMessageContext *context);
 
 void sirc_event_hdr(SircSession *sirc, SircMessage *imsg){
+    GDateTime *time = NULL;
+
+    for (size_t i=0; i<imsg->ntags; i++) {
+        if (!strcmp(imsg->tags[i].key, "time") && imsg->tags[i].value) {
```
Comment (SilverRainZ) [inline src/sirc/sirc_event_hdr.c:45]: A glib version `g_strcmp0` would be better.

File: src/sirc/sirc.c (hunk 1 above)
Comment (progval) [reply src/sirc/sirc.c]: Oh nice, thanks! I'll use `g_autoptr` (which seems to be the right one here)

File: src/sirc/sirc_context.c (hunk 2 above)
Comment (progval) [reply src/sirc/sirc_context.c]: "This function will always succeed unless GLib is still being used after the year 9999."

fair enough

File: src/core/app_ui_event.c [hunk 5]
```diff
@@ -196,13 +204,19 @@ static SrnRet ui_event_quit(SuiBuffer *sui, SuiEvent event, GVariantDict *params
     chat = ctx_get_chat(sui);
     g_return_val_if_fail(srn_server_is_valid(srv), SRN_ERR);
 
+    SircMessageContext *context = sirc_message_context_new(NULL);
```
Comment (SilverRainZ) [inline src/core/app_ui_event.c:207]: We can also use g_autoptr here.

File: src/sirc/sirc.c (hunk 1 above)
Comment (SilverRainZ) [reply src/sirc/sirc.c]: There are other calls to sirc_message_context_free in the code, please check them : )

Comment (progval) [reply src/sirc/sirc.c]: including a couple of memory leaks, oops

**Review by SilverRainZ — APPROVED** [review]
LGTM

**Pull request merged**